callouts/python/extproc/service/network_callout_server.py

```diff
@@ -0,0 +1,364 @@
+
+    self.plaintext_address: tuple[str, int] | None = None
+    if not disable_plaintext:
+      self.plaintext_address = plaintext_address or (default_ip, 8080)
+      if plaintext_port:
+        self.plaintext_address = (self.plaintext_address[0], plaintext_port)
+
+    self.health_check_address: tuple[str, int] | None = None
+    if not combined_health_check:
+      self.health_check_address = health_check_address or (default_ip, 80)
+      if health_check_port:
+        self.health_check_address = (self.health_check_address[0],
+                                     health_check_port)
+
+    def _read_cert_file(path: str | None) -> bytes | None:
+      if path:
+        with open(path, 'rb') as file:
+          return file.read()
+      return None
+
+    self.server_thread_count = server_thread_count
+    self.secure_health_check = secure_health_check
+    # Read cert data.
+    self.private_key = private_key or _read_cert_file(private_key_path)
+    self.cert_chain = cert_chain or _read_cert_file(cert_chain_path)
+
+    if secure_health_check:
+      if not private_key_path:
+        logging.error("Secure health check requires a private_key_path.")
+        return
+      if not cert_chain_path:
+        logging.error("Secure health check requires a cert_chain_path.")
+        return
+      self.health_check_ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)
+      self.health_check_ssl_context.load_cert_chain(certfile=cert_chain_path,
+                                                    keyfile=private_key_path)
+
+    self._callout_server = _GRPCCalloutService(self)
+
+  def run(self) -> None:
+    """Start all requested servers and listen for new connections; blocking."""
+    self._start_servers()
+    self._setup = True
+    try:
+      self._loop_server()
+    except KeyboardInterrupt:
+      logging.info('Server interrupted')
+    finally:
+      self._stop_servers()
+      self._closed = True
+
+  def _start_servers(self) -> None:
+    """Start the requested servers."""
+    if self.health_check_address:
+      self._health_check_server = HTTPServer(self.health_check_address,
+                                             HealthCheckService)
+      protocol = 'HTTP'
+      if self.secure_health_check:
+        protocol = 'HTTPS'
+        self._health_check_server.socket = (
+          self.health_check_ssl_context.wrap_socket(
+            sock=self._health_check_server.socket,))
+
+      logging.info('%s health check server bound to %s.', protocol,
+                   _addr_to_str(self.health_check_address))
+    self._callout_server.start()
+
+  def _stop_servers(self) -> None:
+    """Close the sockets of all servers, and trigger shutdowns."""
+    if self._health_check_server:
+      self._health_check_server.server_close()
+      self._health_check_server.shutdown()
+      logging.info('Health check server stopped.')
+
+    if self._callout_server:
+      self._callout_server.stop()
+
+  def _loop_server(self) -> None:
```

Comment:
Not suggesting a change for this PR, but noticing that there is a fair amount of repeated boilerplate between this and the L7 callout_server.py, maybe a future PR could refactor to share more code.